src/riscv/lib/src/jit/builder/sequence.rs

```diff
@@ -179,6 +205,73 @@ impl<'jit, MC: MemoryConfig> SequenceBuilder<'jit, MC> {
         }
     }
 
+    /// The exit block is used to write the program counter back to the machine core state, as
+    /// well as returning from the JIT function.
+    fn fill_exit_block(&mut self) {
+        self.builder.switch_to_block(self.exit_block);
+
+        // SAFETY: We're declaring the value as a `I64` which has the same representation of
+        // `Address`.
+        let final_program_counter = unsafe {
+            let final_program_counter = self.builder.append_block_param(self.exit_block, I64);
+            Value::<Address>::from_raw(final_program_counter)
+        };
+
+        write_pc(self, final_program_counter);
+
+        let steps_remaining = self.builder.use_var(self.steps_remaining);
+        let max_steps = self.max_steps_param.to_value();
+        let steps = self.builder.ins().isub(max_steps, steps_remaining);
+
+        self.builder.ins().return_(&[steps]);
+    }
+
+    /// IR for calling the interpreted mode fallback function when unable to continue in JIT mode.
+    /// This exit block writes the current PC back to the machine core state before calling the
+    /// interpreter, and returns the total number of steps executed in JIT and interpreted mode.
+    ///
+    /// TODO RV-842: Introduce counter update here for number of calls to fallback to interpreter.
+    fn fill_fallback_exit_block(&mut self) {
+        self.builder.switch_to_block(self.fallback_exit_block);
+        self.builder.set_cold_block(self.fallback_exit_block);
+
+        // Before calling the fallback function, we must commit the current PC to the machine core state.
+
+        // SAFETY: The first parameter of the block is the `current_pc`, which is an `Address`.
+        let current_pc = unsafe {
+            let raw_value = self.builder.block_params(self.fallback_exit_block)[0];
+            Value::<Address>::from_raw(raw_value)
+        };
+
+        write_pc(self, current_pc);
+
+        // SAFETY: The second parameter of the block is the `steps_remaining` value, which is a `usize`.
+        let fallback_max_steps = unsafe {
+            let raw_value = self.builder.block_params(self.fallback_exit_block)[1];
+            Value::<usize>::from_raw(raw_value)
+        };
+
+        // Call the external function to complete the steps in interpreted mode.
```

Comment:
nit - this isn't 'completing' the steps, but rather continuing 'up to' remaining steps